src/Core/Mesh/TriangleMesh.hpp

```diff
@@ -1,38 +1,282 @@
-#include <Core/Math/LinearAlgebra.hpp>
-#include <Core/Mesh/MeshTypes.hpp>
-#include <Core/RaCore.hpp>
-
-namespace Ra {
-namespace Core {
-/// A very basic structure representing a triangle mesh which stores the bare minimum :
-/// vertices, faces and normals. See MeshUtils for geometric functions operating on a mesh.
-struct TriangleMesh {
-    /// Create an empty mesh.
-    inline TriangleMesh() {}
-    /// Copy constructor and assignment operator
-    TriangleMesh( const TriangleMesh& ) = default;
-    TriangleMesh& operator=( const TriangleMesh& ) = default;
-
-    /// Erases all data, making the mesh empty.
-    inline void clear();
-
-    /// Appends another mesh to this one.
-    inline void append( const TriangleMesh& other );
-
-    VectorArray<Vector3> m_vertices;
-    VectorArray<Vector3> m_normals;
-    VectorArray<Triangle> m_triangles;
-
-  public:
-    EIGEN_MAKE_ALIGNED_OPERATOR_NEW
-};
-} // namespace Core
-} // namespace Ra
-
-#include <Core/Mesh/TriangleMesh.inl>
-
-#endif // RADIUMENGINE_TRIANGLEMESH_HPP
+#ifndef RADIUMENGINE_TRIANGLEMESH_HPP
+#define RADIUMENGINE_TRIANGLEMESH_HPP
+
+#include <Core/Containers/VectorArray.hpp>
+#include <Core/Math/LinearAlgebra.hpp>
+#include <Core/Mesh/MeshTypes.hpp>
+#include <Core/RaCore.hpp>
+
+namespace Ra {
+namespace Core {
+
+template <typename T>
+class VertexAttrib;
+
+/// VertexAttribBase is the base class for VertexAttrib.
+/// A VertexAttrib is data linked to Vertices of a mesh
+/// It is also used for rendering, and
+class VertexAttribBase {
+  public:
+    /// attrib name is used to automatic location binding when using shaders.
+    virtual ~VertexAttribBase() {}
+    std::string getName() const { return m_name; }
+    void setName( std::string name ) { m_name = name; }
+    virtual void resize( size_t s ) = 0;
+
+    virtual uint getSize() = 0;
+    virtual int getStride() = 0;
+
+    bool inline operator==( const VertexAttribBase& rhs ) { return m_name == rhs.getName(); }
+
+    template <typename T>
+    inline VertexAttrib<T>& cast() {
+        return static_cast<VertexAttrib<T>&>( *this );
+    }
+
+    template <typename T>
+    inline const VertexAttrib<T>& cast() const {
+        return static_cast<const VertexAttrib<T>&>( *this );
+    }
+
+  private:
+    std::string m_name;
+};
+
+template <typename T>
+class VertexAttrib : public VertexAttribBase {
+  public:
+    using value_type = T;
+    using Container = VectorArray<T>;
+
+    /// resize the container (value_type must have a default ctor).
+    void resize( size_t s ) override { m_data.resize( s ); }
+
+    /// RW acces to container data
+    inline Container& data() { return m_data; }
+
+    /// R only acccess to container data
+    inline const Container& data() const { return m_data; }
+
+    virtual ~VertexAttrib() { m_data.clear(); }
+    uint getSize() override { return Container::Vector::RowsAtCompileTime; }
+    int getStride() override { return sizeof( typename Container::Vector ); }
```

Comment:
@dlyr , what is the purpose of these two functions `getSize` and `getStride` ?

They assume a type `Container::Vector` is defined, but it is not always true (see `VectorArray<Scalar>` specialization [here](https://github.com/STORM-IRIT/Radium-Engine/blob/topological_on_master-v1/src/Core/Containers/VectorArray.hpp#L47).
This currently makes the CI on windows failing (only on windows, as GCC ignore these functions as they are not explicitly used.

Reply:
It's in preparation of the automatic GPU allocation for attribs (but since it's not functionnal yet, we can comment or remove them for now)